src/surface.cpp

```diff
@@ -192,6 +195,59 @@ Surface::reflect(Position r, Direction u) const
   return u -= (2.0 * projection / magnitude) * n;
 }
 
+Direction
+Surface::diffuse_reflect(Position r, Direction u) const
+{
+  // Diffuse reflect direction according to the normal.
+  // cosine distribution 
+  
+  Direction n = normal(r);
+  n = n/n.norm();
+  const double projection = std::max(-1.0, std::min(1.0, n.dot(u)));
+  
+  // sample from inverse function, u=sqrt(rand) since p(u)=2u, so F(u)=u^2 
+  const double cosine = (projection>=0.0) ? 
+                  -std::sqrt(prn()) : std::sqrt(prn());  
+  
+  // handle special case for very large cosine  
+  if(std::fabs(cosine)>1.0001){
+    std::cout<<"cosine="<<"\n";
+    exit(-1);
+  }else if(std::fabs(cosine)>=1.000){
+    return u = cosine * n;
+  }
+
+  // sample azimuthal distribution uniformly 
+  double t1, t2, rr, ss, tt;
+  for(;;){
+    t1 = 2.0 * prn() - 1.0;
+    t2 = 2.0 * prn() - 1.0;  
+    rr = t1 * t1 + t2 * t2;
+    if(rr<=1.0) break;
+  }
+  rr = std::sqrt((1.0 - cosine * cosine)/rr);
+  t1 = t1 * rr;
+  t2 = t2 * rr;
+  
+  // determine the direction off surface relative to norm 
+  if (std::fabs(n[2]<=0.9)) { // for the general normal
+    tt = (std::sqrt(n[0]*n[0]+n[1]*n[1]));
+    ss = 1.0/tt;
+    u[0] = n[0] * cosine + (t1*n[0]*n[2]-t2*n[1])*ss;
+    u[1] = n[1] * cosine + (t1*n[1]*n[2]+t2*n[0])*ss;
+    u[2] = n[2] * cosine - t1*tt;    
+  } else {  // for the normal is almost along with z-axis  
+    tt = (std::sqrt(n[0]*n[0]+n[2]*n[2]));
+    ss = 1.0/tt;   
+    u[0] = n[0] * cosine + (t1*n[0]*n[1]+t2*n[2])*ss;
+    u[1] = n[1] * cosine - t1*tt;
+    u[2] = n[2] * cosine + (t1*n[1]*n[2]-t2*n[0])*ss; 
+  }
```

Comment:
I believe lines 220 up to here can just be replaced by `rotate_angle` from math_functions.cpp